vault/core.go

```diff
@@ -2332,8 +2332,48 @@ func (c *Core) postUnseal(ctx context.Context, ctxCancelFunc context.CancelFunc,
 	// This is intentionally the last block in this function. We want to allow
 	// writes just before allowing client requests, to ensure everything has
 	// been set up properly before any writes can have happened.
-	for _, v := range c.postUnsealFuncs {
-		v()
+	//
+	// Use a small temporary worker pool to run postUnsealFuncs in parallel
+	postUnsealFuncConcurrency := 32
+	if v := os.Getenv("VAULT_POSTUNSEAL_FUNC_CONCURRENCY"); v != "" {
+		pv, err := strconv.Atoi(v)
+		if err != nil || pv < 1 {
+			c.logger.Warn("invalid value for VAULT_POSTUNSEAL_FUNC_CURRENCY, must be a positive integer", "error", err, "value", pv)
+		} else {
+			postUnsealFuncConcurrency = pv
+		}
+	}
+	if postUnsealFuncConcurrency <= 1 {
+		// Out of paranoia, keep the old logic for parallism=1
+		for _, v := range c.postUnsealFuncs {
+			v()
+		}
+	} else {
+		workerChans := make([]chan func(), postUnsealFuncConcurrency)
+		var wg sync.WaitGroup
+		for i := 0; i < postUnsealFuncConcurrency; i++ {
+			wc := make(chan func())
+			workerChans[i] = wc
+			go func() {
+				for {
+					v, more := <-wc
+					if more {
+						wg.Done()
+						v()
+					} else {
+						break
+					}
+				}
+			}()
+		}
+		for i, v := range c.postUnsealFuncs {
+			wg.Add(1)
+			workerChans[i%postUnsealFuncConcurrency] <- v
```

Comment:
I don't know a terrible amount about `c.postUnsealFuncs` but if this variable contains truly independent functions that we can run in parallel, I wonder if it makes more sense to just loop over them and run each one in its own goroutine, vs having a worker pool with a static size.

My main concern with this particular piece is index wraparound when `len(c.postUnsealFuncs) > postUnsealFuncConcurrency`. At that point, you'll be overwriting the contents of the `workerChans` slice, no? Given the default size of 32, maybe the probability of this is low (I don't know what `len(c.postUnsealFuncs)` typically is), but since the size is tunable, it could be anything.

Thoughts?

Reply:
The issue concretely is that we have some customers with 5k+ PKI mounts that need to be migrated when upgrading from <=1.10.x to 1.11+. 

If we run 4k+ PKI mount migration in full parallel, we absolutely slam backend storage and also the Vault node's CPU, in addition to potentially having too many goroutines, thus starving Raft's heatbeats (if its in use). 

Setting a relatively safe upper limit here is good I think. I'd like to maybe see 2x CPU count by default rather than a hard-coded default (just to prevent starvation on nodes with a single CPU core).

Each mount's migration/initialize func yes, is strictly independent, but the combination of work of all of them might be enough to bring down Vault if run in parallel.

Today, we run them strictly serially, which does avoid this problem, but means we strictly stall for IO, then do compute, then stall for more IO, &c.. --- and if each migration only takes 100ms, on 5k+ mounts, we're sitting north of 8 minutes to even unseal, which blows past the default context deadline.

Reply:
Yeah, I'd thought about runtime.NumCPUS(), which I'm back to liking again.